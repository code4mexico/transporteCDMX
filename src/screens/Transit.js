import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import MapView, { Marker } from 'react-native-maps'
import { StatusBar } from 'react-native'
import SplashScreen from 'react-native-splash-screen'
import LoadingOverlay from '../components/LoadingOverlay'
import ErrorOverlay from '../components/ErrorOverlay'
import theme, { sharedStyles } from '../styles'
import storage from '../models/storage'
import { getMetroBusLines } from '../api/endpoints'
import { HTTP_NO_CONTENT, HTTP_SUCCESS } from '../api/request'
import MetrobusLine from '../models/MetrobusLine'

const METROBUS_STORAGE_KEY = '@metrobusLines'
const DEFAULT_MAP_REGION = {
  latitude: 19.42,
  longitude: -99.135,
  latitudeDelta: 0.22,
  longitudeDelta: 0.22,
}

class Transit extends PureComponent {
  state = {
    metrobusLines: null,
    isLoading: true,
    error: false,
  }

  metrobusLines = null

  async componentDidMount() {
    SplashScreen.hide()
    this.metrobusLines = await storage.get(METROBUS_STORAGE_KEY)
    // TODO: We need to ask if ids are prone to change
    if (!this.metrobusLines) {
      this._fetchMetrobusLines()
    } else {
      this.setState({ metrobusLines: JSON.parse(this.metrobusLines), isLoading: false })
    }
  }

  _fetchMetrobusLines = async () => {
    this._metrobusLinesResponseHandler(await getMetroBusLines())
  }

  _metrobusLinesResponseSuccess = async metroBusResponse => {
    if (metroBusResponse?.data?.response) {
      const metrobusLines = metroBusResponse.data.response.map(metrobusLineResponse => {
        return new MetrobusLine(metrobusLineResponse)
      })
      await storage.set(METROBUS_STORAGE_KEY, JSON.stringify(metrobusLines))
      this.setState({ metrobusLines, error: false, isLoading: false })
    } else {
      this.setState({ metrobusLines: null, error: true, isLoading: false })
    }
  }

  _metrobusLinesResponseHandler = metroBusResponse => {
    if (metroBusResponse?.data?.code === HTTP_SUCCESS) {
      this._metrobusLinesResponseSuccess(metroBusResponse)
    } else if (metroBusResponse?.data?.code === HTTP_NO_CONTENT) {
      this.setState({ metrobusLines: [], error: false, isLoading: false })
    } else {
      this.setState({ metrobusLines: null, error: true, isLoading: false })
    }
  }

  _stationPressed = (station, lineColor) => {
    return () => {
      this.props.navigation.navigate('Station Detail', { station, lineColor })
    }
  }

  _renderLinesMarkers = () => {
    return this.state.metrobusLines?.map(line => {
      return line?.stations.map(station => {
        return (
          <Marker
            key={station.id}
            pinColor={line.color}
            coordinate={station.coordinates}
            onPress={this._stationPressed(station, line.color)}
          />
        )
      })
    })
  }

  render() {
    return (
      <>
        <StatusBar barStyle="dark-content" backgroundColor={theme.colors.primary} />
        <MapView initialRegion={DEFAULT_MAP_REGION} style={sharedStyles.flex1}>
          {this._renderLinesMarkers()}
        </MapView>
        <ErrorOverlay showOverlay={this.state.error} onPress={this._fetchMetrobusLines} />
        <LoadingOverlay showOverlay={this.state.isLoading} loaderColor={theme.colors.primary} />
      </>
    )
  }
}

Transit.propTypes = {
  navigation: PropTypes.object,
}

export default Transit
export { METROBUS_STORAGE_KEY }
