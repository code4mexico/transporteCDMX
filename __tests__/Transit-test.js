import React from 'react'
import { shallow } from 'enzyme'
import Transit from '../src/screens/Transit'
import metrobusLine from './fixtures/metrobusLines-fixture'

const metrobusLines = [metrobusLine]

describe('Transit', () => {
  it('it renders correctly while loading', () => {
    const wrapper = shallow(<Transit />)
    expect(wrapper).toMatchSnapshot()
  })

  it('it renders correctly while NOT loading', () => {
    const wrapper = shallow(<Transit />)
    wrapper.setState({ isLoading: false, metrobusLines })
    expect(wrapper).toMatchSnapshot()
  })

  it('it renders correctly with error', () => {
    const wrapper = shallow(<Transit />)
    wrapper.setState({ isLoading: false, error: true })
    expect(wrapper).toMatchSnapshot()
  })
})
