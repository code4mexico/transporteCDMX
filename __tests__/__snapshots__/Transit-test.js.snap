// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transit it renders correctly while NOT loading 1`] = `
<Fragment>
  <StatusBar
    animated={false}
    backgroundColor="#FFF"
    barStyle="dark-content"
    showHideTransition="fade"
  />
  <MapView
    initialRegion={
      Object {
        "latitude": 19.42,
        "latitudeDelta": 0.22,
        "longitude": -99.135,
        "longitudeDelta": 0.22,
      }
    }
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <MapMarker
      coordinate={
        Object {
          "latitude": 19,
          "latitudeDelta": 0.01,
          "longitude": -79,
          "longitudeDelta": 0.01,
        }
      }
      key="7654321"
      onPress={[Function]}
      pinColor="#AAA"
      stopPropagation={false}
    />
    <MapMarker
      coordinate={
        Object {
          "latitude": 19,
          "latitudeDelta": 0.01,
          "longitude": -79,
          "longitudeDelta": 0.01,
        }
      }
      key="7654321"
      onPress={[Function]}
      pinColor="#AAA"
      stopPropagation={false}
    />
  </MapView>
  <Memo(ErrorOverlay)
    onPress={[Function]}
    showOverlay={false}
  />
  <Memo(LoadingOverlay)
    loaderColor="#FFF"
    showOverlay={false}
  />
</Fragment>
`;

exports[`Transit it renders correctly while loading 1`] = `
<Fragment>
  <StatusBar
    animated={false}
    backgroundColor="#FFF"
    barStyle="dark-content"
    showHideTransition="fade"
  />
  <MapView
    initialRegion={
      Object {
        "latitude": 19.42,
        "latitudeDelta": 0.22,
        "longitude": -99.135,
        "longitudeDelta": 0.22,
      }
    }
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <Memo(ErrorOverlay)
    onPress={[Function]}
    showOverlay={false}
  />
  <Memo(LoadingOverlay)
    loaderColor="#FFF"
    showOverlay={true}
  />
</Fragment>
`;

exports[`Transit it renders correctly with error 1`] = `
<Fragment>
  <StatusBar
    animated={false}
    backgroundColor="#FFF"
    barStyle="dark-content"
    showHideTransition="fade"
  />
  <MapView
    initialRegion={
      Object {
        "latitude": 19.42,
        "latitudeDelta": 0.22,
        "longitude": -99.135,
        "longitudeDelta": 0.22,
      }
    }
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <Memo(ErrorOverlay)
    onPress={[Function]}
    showOverlay={true}
  />
  <Memo(LoadingOverlay)
    loaderColor="#FFF"
    showOverlay={false}
  />
</Fragment>
`;
